src/hotspot/share/utilities/formatBuffer.hpp

```diff
@@ -113,7 +113,25 @@ void FormatBuffer<bufsz>::append(const char* format, ...) {
   va_end(argp);
 }
 
-// Used to format messages.
-typedef FormatBuffer<> err_msg;
+template <size_t bufsz = FormatBufferBase::BufferSize>
+class FormatErrBuffer : public FormatBuffer<bufsz> {
+public:
+  inline FormatErrBuffer(const char* format, ...) ATTRIBUTE_PRINTF(2, 3);
+
+  // Dummy overload preventing misuse of err_msg for a string without format arguments.
+  // If compilation fails because of ambiguity between this and real constructor, you
+  // could drop err_msg use at all.
+  inline FormatErrBuffer(const char* msg) { ShouldNotReachHere(); }
```

Comment:
I do not think this is a good idea (apart from it being too complex for a not that serious issue).

The asserts fire, of course, only at runtime. But this function is used usually in some error context, as part of of error reporting. I do not think our tests cover all those paths.

Either somehow make this a compile time error or just leave it as it is. We also could, since this buffer object is used usually as input for vm_exit_during_initialization(), give that function a var-arg overload.

Reply:
Actually, this code prevents a single string argument in compile time. It relies on two constructors to introduce ambiguity in overload resolution that makes C++ compiler complain and abort compilation. Comment above the dummy constructor should clarify that for anyone stepping on compile error.
```
  inline FormatErrBuffer(const char* format, ...) ATTRIBUTE_PRINTF(2, 3);
  inline FormatErrBuffer(const char* msg) { ShouldNotReachHere(); }
```
For sanity check, I've used this sample code https://godbolt.org/z/szs6rE

And the cases that were fixed have been detected by the compiler.

I don't think that this problem is a serious issue as well. But as for me, it is a minor code complexity increase to ensure that the minor problem of extra string copy will never appear again.

Reply:
Oh, now I see it. Smart.

But it occurred to me that someone may want to start a err_msg buffer up with a string literal, only to then add additional content via FormatBuffer::append(). So initializing it with a literal and no arguments may be a valid usecase. 

So I'd still prefer keeping this out. Alternatively, if you like to keep it in, could we just not implement the second constructor? Which should give us linker errors if the static check fails.

Reply:
+1 to leave this undefined to get a linkage error. There are already precedents to do this in Hotspot code, for example

```
  Node(const Node&);            // not defined; linker error to use these
  ...
  // should never be used
  AdapterHandlerEntry();
```

Reply:
This is a good suggestion. Link-time failure is certainly better than runtime one, thanks!

Although err_msg was implemented with FormatBuffer since the beginning [1], I assume it should be an implementation detail. Otherwise there would be no function-like err_msg thing. And at the present, there is only a single case of append to err_msg.

FormatErrBuffer can inherit FormatBuffer privately. It will completely remove err_msg interface relation with FormatBuffer, which is aligned with the most of current uses.

Although I don't clearly understand reasons not to introduce this (beside it may be a bit over-engineered, but without implications to maintainability, I suppose), this may change assumptions about err_msg.

[1] http://hg.openjdk.java.net/jdk10/jdk10/hotspot/rev/f03d0a26bf83#l32.32